### PR TITLE
Add FV tests for "main"

### PR DESCRIPTION
**Pull request by fasaxc**
Move the "main" function to a library package and break it up for ease of mocking/testing.

Builds heavily on #20

### CODE REVIEW

File: pkg/daemon/daemon_test.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+			loggingConfigured = true
+		}
+	})
+
+	It("shouldn't panic when DoEarlyRuntimeSetup is called", func() {
+		d.DoEarlyRuntimeSetup()
+		Expect(earlyLoggingConfigured).To(BeTrue())
+	})
+
+	It("should parse the config file path", func() {
```
Comment (neiljerram) [inline pkg/daemon/daemon_test.go]: Repeat with --config, to check that that works too.

File: pkg/daemon/daemon_test.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+		It("should create the server components", func() {
+			d.CreateServer()
+			Expect(d.SyncerToValidator).ToNot(BeNil())
+			Expect(d.Syncer).ToNot(BeNil())
+			Expect(d.SyncerToValidator).ToNot(BeNil())
+			Expect(d.ValidatorToCache).ToNot(BeNil())
+			Expect(d.Validator).ToNot(BeNil())
+			Expect(d.Cache).ToNot(BeNil())
+			Expect(d.Server).ToNot(BeNil())
```
Comment (neiljerram) [inline pkg/daemon/daemon_test.go]: Split infinitives!  :-)

(I am not being serious.  However, ICYMI, you _can_ also use NotTo.)

File: pkg/daemon/daemon_test.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		})
+
+		It("should start a working server", func() {
+			// Bypass the config validation to tell the server to pick a random port (so we won't clash)
+			d.ConfigParams.ServerPort = syncserver.PortRandom
+			d.CreateServer()
+
+			// Start the server with a context that we can cancel.
+			cxt, cancelFn := context.WithCancel(context.Background())
+			defer cancelFn()
```
Comment (neiljerram) [inline pkg/daemon/daemon_test.go]: So we will end by cancelling the server - but then is there any testing that it exits cleanly?  Should there be?

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+  calico-typha [options]
+
+Options:
+  -c --config-file=<filename>  Config file to load [default: /etc/calico/typha.cfg].
+  --version                    Print the version and exit.
+`
+
+// TyphaDaemon handles the lifecycle of the Typha process.  The main() function of the Typha executable
+// should simply call InitializeAndServeForever() to start the Typha server.  The lifecycle is broken out into
+// serveral individual methods for ease of testing.
```
Comment (neiljerram) [inline pkg/daemon/daemon.go]: "several"

File: pkg/daemon/daemon_test.go (hunk 3 above)
Comment (fasaxc) [inline pkg/daemon/daemon_test.go]: The context is really there just to let the tests tear it down.  Not sure what I can do to improve on that.

**Pull request merged**